Python/Product/PythonTools/PythonTools/TestAdapter/TestContainerDiscoverer.cs

```diff
@@ -0,0 +1,258 @@
+                   new SolutionEventsListener(serviceProvider),
+                    operationState) { }
+
+        internal bool IsProjectKnown(IVsProject project) {
+            var pyProj = project.GetPythonProject();
+            if (pyProj != null) {
+                return _projectInfo.ContainsKey(pyProj);
+            }
+            return false;
+        }
+
+        public TestContainerDiscoverer(IServiceProvider serviceProvider,
+                                       SolutionEventsListener solutionListener,
+                                       IOperationState operationState) {
+            ValidateArg.NotNull(serviceProvider, "serviceProvider");
+            ValidateArg.NotNull(solutionListener, "solutionListener");
+            ValidateArg.NotNull(operationState, "operationState");
+
+            _projectInfo = new Dictionary<PythonProjectNode, ProjectInfo>();
+
+            _serviceProvider = serviceProvider;
+
+            _solutionListener = solutionListener;
+            _solutionListener.ProjectLoaded += OnProjectLoaded;
+            _solutionListener.ProjectUnloading += OnProjectUnloaded;
+            _solutionListener.ProjectClosing += OnProjectUnloaded;
+
+            _firstLoad = true;
+        }
+
+        void IDisposable.Dispose() {
+            if (!_isDisposed) {
+                _isDisposed = true;
+                _solutionListener.Dispose();
+            }
+        }
+
+        public Uri ExecutorUri {
+            get {
+                return _ExecutorUri;
+            }
+        }
+
+        public IEnumerable<ITestContainer> TestContainers {
+            get {
+                if (_firstLoad) {
+                    // Get current solution
+                    var solution = (IVsSolution)_serviceProvider.GetService(typeof(SVsSolution));
+                    // The first time through, we don't know about any loaded
+                    // projects.
+                    _firstLoad = false;
+                    foreach (var project in EnumerateLoadedProjects(solution)) {
+                        OnProjectLoaded(null, new ProjectEventArgs(project));
+                    }
+                    _solutionListener.StartListeningForChanges();
+                }
+
+                return _projectInfo.Values.SelectMany(x => x._containers).Select(x => x.Value);
+            }
+        }
+
+        public TestContainer GetTestContainer(PythonProjectNode project, string path) {
+            ProjectInfo projectInfo;
+            if (_projectInfo.TryGetValue(project, out projectInfo)) {
+                var analysis = project.GetAnalyzer().GetAnalysisEntryFromPath(path);
+                TestContainer container;
+                if (analysis != null && projectInfo._containers.TryGetValue(analysis, out container)) {
+                    return container;
+                }
+            }
+
+            return null;
+        }
+
+        private static IEnumerable<IVsProject> EnumerateLoadedProjects(IVsSolution solution) {
```

Comment:
There's already an extension method on `IVsSolution` for this (in `PythonTools` project), but if we move this class back to its own assembly then I guess this has to stay.